wagtail/admin/views/pages/choose_parent.py

```diff
@@ -18,6 +20,66 @@ class ChooseParentView(WagtailAdminTemplateMixin, FormView):
     index_url_name = None
     page_title = gettext_lazy("Choose parent")
 
+    def get_valid_parent_pages(self, user):
+        """
+        Identifies possible parent pages for the current user by first looking
+        at allowed_parent_page_models() on self.model to limit options to the
+        correct type of page, then checking permissions on those individual
+        pages to make sure we have permission to add a subpage to it.
+        """
+        # Get queryset of pages where this page type can be added
+        allowed_parent_page_content_types = list(
+            ContentType.objects.get_for_models(
+                *self.model.allowed_parent_page_models()
+            ).values()
+        )
+        allowed_parent_pages = Page.objects.filter(
+            content_type__in=allowed_parent_page_content_types
+        )
+
+        # Get queryset of pages where the user has permission to add subpages
+        if user.is_superuser:
+            pages_where_user_can_add = Page.objects.all()
+        else:
+            pages_where_user_can_add = Page.objects.none()
+
+            perms = {
+                perm
+                for perm in page_permission_policy.get_cached_permissions_for_user(user)
+                if perm.permission.codename == "add_page"
+            }
+
+            for perm in perms:
+                # user has add permission on any subpage of perm.page
+                # (including perm.page itself)
+                pages_where_user_can_add |= Page.objects.descendant_of(
+                    perm.page, inclusive=True
+                )
+
+        # Combine them
+        return allowed_parent_pages & pages_where_user_can_add
+
+    def get(self, request, *args, **kwargs):
+        parents = self.get_valid_parent_pages(request.user)
+
+        # Only fetch the IDs for the first two parents to check if there's only
+        # one valid parent, and if so, redirect to the add page view with that
+        # parent pre-selected
+        parent_ids = parents.values_list("pk", flat=True)[:2]
+        if len(parent_ids) == 1:
```

Comment:
I changed this to use the single query instead, since I _think_ it should be more efficient compared to doing a count on the queryset and then fetching the first one. I'm assuming that performance-wise, fetching one result isn't all that different to fetching two. Plus, doing a count _might_ be expensive (and we're not interested in the actual count itself).

Reply:
@laymonage I think the other way was a bit better readable (thats why I picked it instead of this), but I totally understand that it makes sense performance wise.